File: elixir/lib/2024/03_mining_maestro/ec202403.ex
defmodule EverybodyCodes2024.Day03 do
  @moduledoc """
  Everybody Codes 2024, day 3: Mining Maestro
  """
  require EverybodyCodes

  @neighbors_adjacent [{-1, 0}, {0, -1}, {0, 1}, {1, 0}]
  @neighbors_diagonal [{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}]

  @doc """
  Solve part 1
  """
  def part1(puzzle_input) do
    puzzle_input |> parse_grid() |> count_blocks(@neighbors_adjacent)
  end

  @doc """
  Solve part 2
  """
  def part2(puzzle_input) do
    puzzle_input |> parse_grid() |> count_blocks(@neighbors_adjacent)
  end

  @doc """
  Solve part 3
  """
  def part3(puzzle_input) do
    puzzle_input |> parse_grid() |> count_blocks(@neighbors_diagonal)
  end

  @doc """
  Parse input into a mapset of coordinates

  ## Example

      iex> parse_grid("...#\\n#.##\\n##..")
      MapSet.new([{0, 3}, {1, 0}, {1, 2}, {1, 3}, {2, 0}, {2, 1}])
  """
  def parse_grid(grid) do
    for {line, row} <- grid |> String.split("\n", trim: true) |> Enum.with_index(),
        {char, col} <- line |> String.to_charlist() |> Enum.with_index(),
        char == ?#,
        do: {row, col},
        into: MapSet.new()
  end

  @doc """
  Count the number of blocks that can be dug out, based on the given neighbors

  ## Example

      iex> count_blocks(
      ...>     MapSet.new([{0, 1}, {1, 0}, {1, 1}, {1, 2}, {2, 1}]),
      ...>     [{-1, 0}, {0, -1}, {0, 1}, {1, 0}]
      ...> )
      6
  """
  def count_blocks(grid, neighbors) do
    if Enum.empty?(grid),
      do: 0,
      else: Enum.count(grid) + count_blocks(grid |> shrink(neighbors), neighbors)
  end

  @doc """
  Shrink a grid to only include those blocks with neighbors on all sides

  ## Example

      iex> shrink(
      ...>     MapSet.new([{0, 1}, {1, 0}, {1, 1}, {1, 2}, {2, 1}]),
      ...>     [{-1, 0}, {0, -1}, {0, 1}, {1, 0}]
      ...> )
      MapSet.new([{1, 1}])
  """
  def shrink(grid, neighbors) do
    grid
    |> Enum.filter(fn {row, col} ->
      Enum.all?(neighbors, fn {drow, dcol} -> {row + drow, col + dcol} in grid end)
    end)
    |> Enum.into(MapSet.new())
  end

  def main(path1, path2, path3) do
    EverybodyCodes.solve(&part1/1, &part2/1, &part3/1, path1, path2, path3)
  end
end

File: python/src/2024_the-kingdom-of-algorithmia/03_mining-maestro/ec202403.py
"""Everybody Codes quest 3, 2024: Mining Maestro."""

from collections.abc import Callable

Point = tuple[int, int]


def parse_data(puzzle_input: str) -> set[Point]:
    """Parse map into set of coordinates"""
    return {
        (row, col)
        for row, line in enumerate(puzzle_input.split("\n"))
        for col, char in enumerate(line)
        if char == "#"
    }


def part1(puzzle_input: str) -> int:
    """Solve part 1."""
    return solve(puzzle_input, neighbors_4)


def part2(puzzle_input: str) -> int:
    """Solve part 2."""
    return solve(puzzle_input, neighbors_4)


def part3(puzzle_input: str) -> int:
    """Solve part 3."""
    return solve(puzzle_input, neighbors_8)


def solve(puzzle_input: str, neighbors: Callable[[Point], set[Point]]) -> int:
    """Solve the puzzle for the given number of neighbors"""
    coords = parse_data(puzzle_input)
    return count_blocks(coords, neighbors)


def count_blocks(coords: set[Point], neighbors: Callable[[Point], set[Point]]) -> int:
    """Count the number of blocks that can be dug on the given map.

    ## Example:

        .#.        .1.        .1.
        ###   ->   111   ->   121
        .#.        .1.        .1.

    >>> count_blocks({(0, 1), (1, 0), (1, 1), (1, 2), (2, 1)}, neighbors_4)
    6
    """
    if not coords:
        return 0
    return len(coords) + count_blocks(
        {coord for coord in coords if not neighbors(coord) - coords}, neighbors
    )


def neighbors_4(point: Point) -> set[Point]:
    """List neighbors of the given point, adjacent only.

    ## Example

    >>> sorted(neighbors_4((2, 5)))
    [(1, 5), (2, 4), (2, 6), (3, 5)]
    """
    row, col = point
    return {(row - 1, col), (row, col - 1), (row, col + 1), (row + 1, col)}


def neighbors_8(point: Point) -> set[Point]:
    """List neighbors of the given point, including diagonally.

    ## Example

    >>> sorted(neighbors_8((2, 5)))
    [(1, 4), (1, 5), (1, 6), (2, 4), (2, 6), (3, 4), (3, 5), (3, 6)]
    """
    row, col = point
    return {
        (row - 1, col - 1),
        (row - 1, col),
        (row - 1, col + 1),
        (row, col - 1),
        (row, col + 1),
        (row + 1, col - 1),
        (row + 1, col),
        (row + 1, col + 1),
    }
